homeassistant/components/opentherm_gw/__init__.py

```diff
@@ -19,340 +20,375 @@
-})
 
 CLIMATE_SCHEMA = vol.Schema({
-    vol.Optional(CONF_NAME, default="OpenTherm Gateway"): cv.string,
     vol.Optional(CONF_PRECISION): vol.In([PRECISION_TENTHS, PRECISION_HALVES,
                                           PRECISION_WHOLE]),
     vol.Optional(CONF_FLOOR_TEMP, default=False): cv.boolean,
 })
 
 CONFIG_SCHEMA = vol.Schema({
-    DOMAIN: vol.Schema({
+    DOMAIN: cv.schema_with_slug_keys({
         vol.Required(CONF_DEVICE): cv.string,
         vol.Optional(CONF_CLIMATE, default={}): CLIMATE_SCHEMA,
-        vol.Optional(CONF_MONITORED_VARIABLES, default=[]): vol.All(
-            cv.ensure_list, [cv.string]),
+        vol.Optional(CONF_MONITORED_VARIABLES, default=[]): [cv.string],
+        vol.Optional(CONF_NAME): cv.string,
     }),
 }, extra=vol.ALLOW_EXTRA)
 
 
 async def async_setup(hass, config):
     """Set up the OpenTherm Gateway component."""
-    import pyotgw
     conf = config[DOMAIN]
-    gateway = pyotgw.pyotgw()
-    monitored_vars = conf.get(CONF_MONITORED_VARIABLES)
-    hass.data[DATA_OPENTHERM_GW] = {
-        DATA_DEVICE: gateway,
-        DATA_GW_VARS: pyotgw.vars,
-        DATA_LATEST_STATUS: {}
-    }
-    hass.async_create_task(register_services(hass, gateway))
-    hass.async_create_task(async_load_platform(
-        hass, 'climate', DOMAIN, conf.get(CONF_CLIMATE), config))
-    if monitored_vars:
-        hass.async_create_task(setup_monitored_vars(
-            hass, config, monitored_vars))
-    # Schedule directly on the loop to avoid blocking HA startup.
-    hass.loop.create_task(
-        connect_and_subscribe(hass, conf[CONF_DEVICE], gateway))
+    hass.data[DATA_OPENTHERM_GW] = {DATA_GATEWAYS: {}}
+    for gw_id, cfg in conf.items():
+        gateway = OpenThermGatewayDevice(hass, gw_id, cfg, config)
+        hass.data[DATA_OPENTHERM_GW][DATA_GATEWAYS][gw_id] = gateway
+    hass.async_create_task(register_services(hass))
     return True
 
 
-async def connect_and_subscribe(hass, device_path, gateway):
-    """Connect to serial device and subscribe report handler."""
-    await gateway.connect(hass.loop, device_path)
-    _LOGGER.debug("Connected to OpenTherm Gateway at %s", device_path)
-
-    async def cleanup(event):
-        """Reset overrides on the gateway."""
-        await gateway.set_control_setpoint(0)
-        await gateway.set_max_relative_mod('-')
-    hass.bus.async_listen(EVENT_HOMEASSISTANT_STOP, cleanup)
-
-    async def handle_report(status):
-        """Handle reports from the OpenTherm Gateway."""
-        _LOGGER.debug("Received report: %s", status)
-        hass.data[DATA_OPENTHERM_GW][DATA_LATEST_STATUS] = status
-        async_dispatcher_send(hass, SIGNAL_OPENTHERM_GW_UPDATE, status)
-    gateway.subscribe(handle_report)
-
-
-async def register_services(hass, gateway):
+async def register_services(hass):
     """Register services for the component."""
-    gw_vars = hass.data[DATA_OPENTHERM_GW][DATA_GW_VARS]
+    import pyotgw.vars as gw_vars
```

Comment:
I thought dependency imports were only allowed inside functions that use them? Or does that rule expand to modules as well?

Reply:
This changed after we introduced manifest.json. We can now import 3rd party modules and packages at the top of the module. This can make some more complicated structures, that were required before, obsolete.